File: examples/sol/merkle-proof/MerkleProofDemo.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract } from 'runar-testing';
import { createHash } from 'node:crypto';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'MerkleProofDemo.runar.sol'), 'utf8');
const FILE_NAME = 'MerkleProofDemo.runar.sol';

// ---------------------------------------------------------------------------
// Merkle tree helpers (for building test fixtures)
// ---------------------------------------------------------------------------

function sha256(hex: string): string {
  return createHash('sha256').update(Buffer.from(hex, 'hex')).digest('hex');
}

function hash256(hex: string): string {
  return sha256(sha256(hex));
}

/**
 * Build a depth-4 Merkle tree (16 leaves) and return the root + proof for a given leaf index.
 */
function buildMerkleTree(leaves: string[], hashFn: (h: string) => string): {
  root: string;
  getProof: (index: number) => { proof: string; leaf: string };
} {
  let level = [...leaves];
  const layers: string[][] = [level];

  while (level.length > 1) {
    const next: string[] = [];
    for (let i = 0; i < level.length; i += 2) {
      next.push(hashFn(level[i]! + level[i + 1]!));
    }
    level = next;
    layers.push(level);
  }

  const root = level[0]!;

  return {
    root,
    getProof(index: number) {
      const siblings: string[] = [];
      let idx = index;
      for (let d = 0; d < layers.length - 1; d++) {
        const siblingIdx = idx ^ 1;
        siblings.push(layers[d]![siblingIdx]!);
        idx = idx >> 1;
      }
      return {
        proof: siblings.join(''),
        leaf: leaves[index]!,
      };
    },
  };
}

describe('MerkleProofDemo (Solidity)', () => {
  // Create 16 leaves (32-byte hashes)
  const leaves: string[] = [];
  for (let i = 0; i < 16; i++) {
    leaves.push(sha256(Buffer.from([i]).toString('hex')));
  }

  describe('verifySha256 (merkleRootSha256, depth=4)', () => {
    const tree = buildMerkleTree(leaves, sha256);

    it('verifies leaf at index 0', () => {
      const { proof, leaf } = tree.getProof(0);
      const c = TestContract.fromSource(source, { expectedRoot: tree.root }, FILE_NAME);
      const r = c.call('verifySha256', { leaf, proof, index: 0n });
      expect(r.success).toBe(true);
    });

    it('verifies leaf at index 7', () => {
      const { proof, leaf } = tree.getProof(7);
      const c = TestContract.fromSource(source, { expectedRoot: tree.root }, FILE_NAME);
      const r = c.call('verifySha256', { leaf, proof, index: 7n });
      expect(r.success).toBe(true);
    });

    it('verifies leaf at index 15', () => {
      const { proof, leaf } = tree.getProof(15);
      const c = TestContract.fromSource(source, { expectedRoot: tree.root }, FILE_NAME);
      const r = c.call('verifySha256', { leaf, proof, index: 15n });
      expect(r.success).toBe(true);
    });

    it('rejects wrong leaf', () => {
      const { proof } = tree.getProof(0);
      const wrongLeaf = sha256('ff');
      const c = TestContract.fromSource(source, { expectedRoot: tree.root }, FILE_NAME);
      const r = c.call('verifySha256', { leaf: wrongLeaf, proof, index: 0n });
      expect(r.success).toBe(false);
    });

    it('rejects wrong index', () => {
      const { proof, leaf } = tree.getProof(0);
      const c = TestContract.fromSource(source, { expectedRoot: tree.root }, FILE_NAME);
      const r = c.call('verifySha256', { leaf, proof, index: 1n });
      expect(r.success).toBe(false);
    });
  });

  describe('verifyHash256 (merkleRootHash256, depth=4)', () => {
    const tree = buildMerkleTree(leaves, hash256);

    it('verifies leaf at index 0', () => {
      const { proof, leaf } = tree.getProof(0);
      const c = TestContract.fromSource(source, { expectedRoot: tree.root }, FILE_NAME);
      const r = c.call('verifyHash256', { leaf, proof, index: 0n });
      expect(r.success).toBe(true);
    });

    it('verifies leaf at index 10', () => {
      const { proof, leaf } = tree.getProof(10);
      const c = TestContract.fromSource(source, { expectedRoot: tree.root }, FILE_NAME);
      const r = c.call('verifyHash256', { leaf, proof, index: 10n });
      expect(r.success).toBe(true);
    });
  });
});

File: examples/zig/merkle-proof/MerkleProofDemo_test.zig
const std = @import("std");

const root = @import("../examples_test.zig");
const runar = @import("runar");
const MerkleProofDemo = @import("MerkleProofDemo.runar.zig").MerkleProofDemo;

fn contractPath(comptime basename: []const u8) []const u8 {
    return "merkle-proof/" ++ basename;
}

fn runCompileChecks(comptime basename: []const u8) !void {
    try root.runar.compileCheckSource(std.testing.allocator, @embedFile(basename), basename);
    try root.runar.compileCheckFile(std.testing.allocator, contractPath(basename));
}

test "compile-check MerkleProofDemo.runar.zig" {
    try runCompileChecks("MerkleProofDemo.runar.zig");
}

// Build a depth-4 SHA-256 Merkle tree from 16 leaves and verify proofs.
test "MerkleProofDemo verifySha256 accepts valid proof at index 0" {
    const leaves = buildLeaves();
    const tree = buildTree(leaves, sha256Fn);
    const proof_data = getProof(tree.layers, leaves, 0);

    const contract = MerkleProofDemo.init(tree.root);
    contract.verifySha256(leaves[0], proof_data.proof, 0);
}

test "MerkleProofDemo verifySha256 accepts valid proof at index 7" {
    const leaves = buildLeaves();
    const tree = buildTree(leaves, sha256Fn);
    const proof_data = getProof(tree.layers, leaves, 7);

    const contract = MerkleProofDemo.init(tree.root);
    contract.verifySha256(leaves[7], proof_data.proof, 7);
}

test "MerkleProofDemo verifyHash256 accepts valid proof at index 0" {
    const leaves = buildLeaves();
    const tree = buildTree(leaves, hash256Fn);
    const proof_data = getProof(tree.layers, leaves, 0);

    const contract = MerkleProofDemo.init(tree.root);
    contract.verifyHash256(leaves[0], proof_data.proof, 0);
}

// --- Helpers ---

fn sha256Fn(data: runar.ByteString) runar.ByteString {
    return runar.sha256(data);
}

fn hash256Fn(data: runar.ByteString) runar.ByteString {
    return runar.hash256(data);
}

const TreeResult = struct {
    root: runar.ByteString,
    layers: [5][16]runar.ByteString,
};

const ProofResult = struct {
    proof: runar.ByteString,
};

fn buildLeaves() [16]runar.ByteString {
    var leaves: [16]runar.ByteString = undefined;
    for (0..16) |i| {
        const byte = [1]u8{@intCast(i)};
        leaves[i] = runar.sha256(&byte);
    }
    return leaves;
}

fn buildTree(leaves: [16]runar.ByteString, hashFn: fn (runar.ByteString) runar.ByteString) TreeResult {
    var layers: [5][16]runar.ByteString = undefined;
    for (0..16) |i| {
        layers[0][i] = leaves[i];
    }

    var level_size: usize = 16;
    var level: usize = 0;
    while (level_size > 1) {
        const next_size = level_size / 2;
        for (0..next_size) |i| {
            layers[level + 1][i] = hashFn(runar.bytesConcat(layers[level][i * 2], layers[level][i * 2 + 1]));
        }
        level += 1;
        level_size = next_size;
    }

    return .{ .root = layers[4][0], .layers = layers };
}

fn getProof(layers: [5][16]runar.ByteString, leaves: [16]runar.ByteString, index: usize) ProofResult {
    _ = leaves;
    var siblings: [4]runar.ByteString = undefined;
    var idx = index;
    for (0..4) |d| {
        const sibling_idx = idx ^ 1;
        siblings[d] = layers[d][sibling_idx];
        idx = idx >> 1;
    }

    var proof = siblings[0];
    for (1..4) |i| {
        proof = runar.bytesConcat(proof, siblings[i]);
    }

    return .{ .proof = proof };
}
